atsign-foundation/at_server
# feat: Persistence impl refactoring

Persistence :
**- What I did**
- uptake persistence spec changes for put/create methods and refactoring related to the changes

**- How I did it**
- in hive_keystore.dart, modified impl of put/create methods. version and createdAt/updatedAt metadata is set just before putting/creating the key. Remove putAll method.
- AtMetadataBuilder was having inconsistent logic for setting metadata fields. Removed the code from persistence secondary
- in hive_keystore.dart, removed AtMetadataBuilder. metadata passed to the method will be used without modification(except version and updatedAt)
- in hive_keystore_helper.dart -> prepareDataForKeystoreOperation, removed AtMetadataBuilder. this will be the order of priority for setting newMetaData
    1. passed in metadata
    2. metadata from passed value
    3. existing metadata
-   in hive_keystore_helper.dart -> prepareDataForKeystoreOperation, derived values from ttl, ttb and ttr are computed
- at_metadata_.dart. removed AtMetadataBuilder.
Changes to tests:
- in at_metadata_test.dart, removed updatedAt check from default field for new key. For new key only createdAt is needed. updatedAt will be set only on subsequent update to value/metadata for that key. Removed test for putAll

**- How to verify it**
- unit, functional and end2endtests should pass



secondary server :
**- What I did**
 - uptake persistence spec and persistence refactoring changes
**- How I did it**
- Moved AtMetadataBuilder from persistence to server and refactored the code. Currently new AtMetadataBuilder is used in UpdateMetaVerbHandler
Changes to tests:
- in sync_unit_test.dart removed updatedAt check for new key creation
**- How to verify it**
- unit, functional and end2endtests should pass


Comments/clarifications/TODO:
* in reset_handling_test.dart, I am not sure of the logic behind cached data createdAt. Temporarily changed the check for the test to pass
* in update_verb_test.dart, "update verb tests with metadata". commented a check since createdAt will be part of commons metadata in new changes. Old code returns null for createdAt and updateBuilder.atKey.metadata also doesn't have createdAt. so the check passes in trunk but fails with latest changes.
* in persistence spec, commons metadata is used as param in put/create methods. Ideally in server layer, commons metadata should be used (e.g verb handlers, metadata builder etc.,) So the conversion from commons metadata to hive metadata happens only once in persistence layer. Currently commons metadata and hive metadata are both used in server. so there is unwanted conversion between commons --> hive metadata and vice versa. Another advantage of using commons metadata only in server, will enable us to replace keystore impl easily in the future. I didn't attempt this refactoring yet.
* some of the logic to set metadata still exists in HiveKeystore --> prepareDataForKeystoreOperation. Ideally all logic to be inside AtMetadataBuilder.Both update verb handler and update meta verb handler should call AtMetadataBuilder without having any major logic in verb handlers/persistence for metadata setting/unsetting.
* I will separate the PRs after addressing review comments, since it is easy to run server tests in local after making changes in persistence with single branch
